### PR TITLE
Avoid ledger copy and mutation in `Sparse_ledger.of_ledger_subset_exn`

### PR DESCRIPTION
**Pull request by mrmr1993**
This PR builds upon #14586, removing the `Ledger.copy` and `Ledger.create_empty_exn` calls from `Sparse_ledger.of_ledger_subset_exn`. This function takes a significant amount of time in staged ledger diff application, and so this PR should further improve performance there.

Checklist:

- [ ] Dependency versions are unchanged
  - Notify Velocity team if dependencies must change in CI
- [ ] Modified the current draft of release notes with details on what is completed or incomplete within this project
- [ ] Document code purpose, how to use it
  - Mention expected invariants, implicit constraints
- [ ] Tests were added for the new behavior
  - Document test purpose, significance of failures
  - Test names should reflect their purpose
- [ ] All tests pass (CI will check this if you didn't)
- [ ] Serialized types are in stable-versioned modules
- [ ] Does this close issues? List them

### CODE REVIEW

**Comment by mrmr1993** [comment]
!ci-build-me

**Comment by mrmr1993** [comment]
!ci-nightly-me

**Comment by mrmr1993** [comment]
!ci-build-me

**Comment by sebastiencs** [comment]
From what I understand, `Ledger.create_empty_exn` is slow because it indirectly calls `Masking_merkle_tree.set`, which re-compute all hashes for each new accounts:
https://github.com/MinaProtocol/mina/blob/55d20405f195b3f8e4d6603e216798191472d707/src/lib/merkle_mask/masking_merkle_tree.ml#L318C1-L327C38

There are multiple PRs opened to optimize transaction applications by avoiding access to disk, but I don't think that the slowness comes from those DB access.

The slowest part is that every modification to a `Mask` re-computes all hashes.
A solution would be to delay re-hashing a `Mask` until one of the method `merkle_root` or `merkle_path` is called.
In the case of `Sparse_ledger.of_ledger_subset_exn`, with new accounts, no hashing is even required since those are empty accounts.

**Comment by mrmr1993** [comment]
> From what I understand, `Ledger.create_empty_exn` is slow because it indirectly calls `Masking_merkle_tree.set`, which re-compute all hashes for each new accounts:

Indeed, we already knew and were working on a fix. To keep the PRs small and testing more modular, it's in its own PR here: https://github.com/MinaProtocol/mina/pull/14594

**Comment by mrmr1993** [comment]
!ci-build-me

File: src/lib/mina_ledger/sparse_ledger.ml [hunk 1]
```diff
@@ -7,33 +7,77 @@ let of_ledger_root ledger =
...
+             (new_location :: merkle_path_locations)
+         else merkle_path_locations
+       in
+       add_locations )
+        num_new_accounts
+        (Ledger.last_filled oledger)
+        non_empty_locations
+    in
+    let merkle_paths = Ledger.merkle_path_batch oledger merkle_path_locations in
+    (let rec pop_empties num_empties merkle_paths locations acc =
```
Comment (ghost-not-in-the-shell) [inline src/lib/mina_ledger/sparse_ledger.ml:46]: Can't we just do a list.split here? We don't use the `location` of empty accounts later.
I think we could just have `List.split merkle_paths num_new_accounts` here.

**Review by ghost-not-in-the-shell — APPROVED** [review]

File: src/lib/mina_ledger/sparse_ledger.ml (hunk 1 above)
Comment (georgeee) [reply src/lib/mina_ledger/sparse_ledger.ml:46]: @ghost-not-in-the-shell do you mind implementing this change and committing to this PR?

**Comment by ghost-not-in-the-shell** [comment]
!ci-build-me

**Comment by deepthiskumar** [comment]
!ci-nightly-me

**Comment by deepthiskumar** [comment]
!ci-build-me

**Comment by deepthiskumar** [comment]
!ci-nightly-me

**Comment by deepthiskumar** [comment]
!ci-nightly-me

**Pull request merged**